aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+                owner: creator_address,
+                allow_ungated_transfer: true,
+                transfer_events: event::new_event_handle(transfer_events_guid),
+            },
+        );
+        CreatorRef { self: id, can_delete }
+    }
+
+    // Creation helpers
+
+    public fun disallow_ungated_transfer(ref: &CreatorRef) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        object.allow_ungated_transfer = false;
+    }
+
+    public fun generate_delete_ref(ref: &CreatorRef): DeleteRef {
+        assert!(ref.can_delete, error::permission_denied(ECANNOT_DELETE));
+        DeleteRef { self: ref.self }
+    }
+
+    /// Create a signer for the CreatorRef
+    public fun generate_signer(ref: &CreatorRef): signer {
+        create_signer(ref.self.inner)
+    }
+
+    public fun object_id_from_creator_ref(ref: &CreatorRef): ObjectId {
+        ref.self
+    }
+
+    // Signer please...
+
+    /// Create a guid for the object, typically used for events
+    public fun create_guid(object: &signer): guid::GUID acquires Object {
+        let addr = signer::address_of(object);
+        let object_data = borrow_global_mut<Object>(addr);
+        guid::create(addr, &mut object_data.guid_creation_num)
+    }
+
+    /// Generate a new event handle
+    public fun new_event_handle<T: drop + store>(
+        object: &signer,
+    ): event::EventHandle<T> acquires Object {
+        event::new_event_handle(create_guid(object))
+    }
+
+    // Deletion helpers
+
+    public fun delete_ref_id(ref: &DeleteRef): ObjectId {
+        ref.self
+    }
+
+    public fun delete(ref: DeleteRef) acquires Object {
+        let object = move_from<Object>(ref.self.inner);
+        let Object {
+            guid_creation_num: _,
+            owner: _,
+            allow_ungated_transfer: _,
+            transfer_events,
+        } = object;
+        event::destroy_handle(transfer_events);
+    }
+
+    // Extension helpers
+
+    public fun generate_extend_ref(ref: &CreatorRef): ExtendRef {
+        ExtendRef { self: ref.self }
+    }
+
+    /// Create a signer for the ExtendRef
+    public fun generate_signer_for_extending(ref: &ExtendRef): signer {
+        create_signer(ref.self.inner)
+    }
+
+    // Transfer functionality
+
+    public fun generate_transfer_ref(ref: &CreatorRef): TransferRef {
+        TransferRef { self: ref.self }
+    }
+
+    public fun generate_linear_transfer_ref(ref: TransferRef): LinearTransferRef {
+        LinearTransferRef { self: ref.self }
+    }
+
+    public fun transfer_with_ref(ref: LinearTransferRef, to: address) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        event::emit_event(
+            &mut object.transfer_events,
+            TransferEvent {
+                object_id: ref.self,
+                from: object.owner,
+                to,
+            },
+        );
+        object.owner = to;
+    }
+
+    public entry fun entry_transfer(
```

Comment:
Do we have this already as a convention? If not I suggest a better naming, perhaps `call_transfer`?